File: lib/aabb.dart
import 'dart:typed_data';

import 'package:flutter_rust_bridge/flutter_rust_bridge.dart';

import 'ffi.dart' if (dart.library.html) 'ffi_web.dart';

class AABB {
  static final Finalizer<Index> _finalizer =
      Finalizer((aabb) => api.aabbDropBulk(aabbIds: [aabb]));

  final int index;
  final int generation;

  Index getIndex() {
    return Index(index: index.toInt(), generation: generation.toInt());
  }

  AABB(this.index, this.generation) {
    _finalizer.attach(
        this, Index(index: index.toInt(), generation: generation.toInt()));
  }

  // Factories
  factory AABB.fromIndex(Index index) {
    return AABB(index.index, index.generation);
  }

  factory AABB.minMax(double minX, double minY, double maxX, double maxY) {
    final id = api.aabbNew(minX: minX, minY: minY, maxX: maxX, maxY: maxY);
    return AABB(id.index, id.generation);
  }

  factory AABB.fromList(List<double> minMax) {
    assert(minMax.length == 4);
    return AABB.minMax(minMax[0], minMax[1], minMax[2], minMax[3]);
  }

  factory AABB.fromXYWH(double x, double y, double width, double height) {
    return AABB.minMax(x, y, x + width, y + height);
  }

  factory AABB.fromXYWHList(List<double> xywh) {
    assert(xywh.length == 4);
    return AABB.fromXYWH(xywh[0], xywh[1], xywh[2], xywh[3]);
  }

  static List<AABB> minMaxRawBulk(Float64List points) {
    const isWeb = kIsWeb;
    final ids = api.aabbNewBulk(points: points);
    final len = ids.length / 8;
    final bytes = ByteData.view(ids.buffer);
    final aabbs = <AABB>[];
    for (var i = 0; i < len; i += 2) {
      final offsetIndex = i * 8;
      final offsetGeneration = (i + 1) * 8;
      if (isWeb) {
        // Work around by getting two Uint32 instead of one Uint64
        // and then combining them into one Uint64
        final idIndexLow = bytes.getUint32(offsetIndex, Endian.host);
        final idIndexHigh = bytes.getUint32(offsetIndex + 4, Endian.host);
        final idGenerationLow = bytes.getUint32(offsetGeneration, Endian.host);
        final idGenerationHigh =
            bytes.getUint32(offsetGeneration + 4, Endian.host);
        final idIndex = (idIndexHigh << 32) + idIndexLow;
        final idGeneration = (idGenerationHigh << 32) + idGenerationLow;
        aabbs.add(AABB(idIndex, idGeneration));
      } else {
        final idIndex = bytes.getUint64(offsetIndex, Endian.host);
        final idGeneration = bytes.getUint64(offsetGeneration, Endian.host);
        aabbs.add(AABB(idIndex, idGeneration));
      }
    }
    return aabbs;
  }

  static List<AABB> minMaxBulk(List<double> minXs, List<double> minYs,
      List<double> maxXs, List<double> maxYs) {
    final points = <double>[];
    for (var i = 0; i < minXs.length; i++) {
      points.add(minXs[i]);
      points.add(minYs[i]);
      points.add(maxXs[i]);
      points.add(maxYs[i]);
    }
    return minMaxRawBulk(Float64List.fromList(points));
  }

  static List<AABB> fromListBulk(List<List<double>> minMaxs) {
    final minXs = <double>[];
    final minYs = <double>[];
    final maxXs = <double>[];
    final maxYs = <double>[];
    for (final minMax in minMaxs) {
      assert(minMax.length == 4);
      minXs.add(minMax[0]);
      minYs.add(minMax[1]);
      maxXs.add(minMax[2]);
      maxYs.add(minMax[3]);
    }
    return minMaxBulk(minXs, minYs, maxXs, maxYs);
  }

  static List<AABB> fromXYWHBulk(List<double> xs, List<double> ys,
      List<double> widths, List<double> heights) {
    assert(xs.length == ys.length);
    assert(xs.length == widths.length);
    assert(xs.length == heights.length);
    final minXs = <double>[];
    final minYs = <double>[];
    final maxXs = <double>[];
    final maxYs = <double>[];
    for (var i = 0; i < xs.length; ++i) {
      final x = xs[i];
      final y = ys[i];
      final width = widths[i];
      final height = heights[i];
      minXs.add(x);
      minYs.add(y);
      maxXs.add(x + width);
      maxYs.add(y + height);
    }
    return minMaxBulk(minXs, minYs, maxXs, maxYs);
  }

  static List<bool> dropBulk(List<AABB> aabbs) {
    final ids = aabbs.map((aabb) => aabb.getIndex()).toList();
    return api.aabbDropBulk(aabbIds: ids).map((i) => i == 1).toList();
  }

  // Properties
  List<double> get min {
    final min = api.aabbMin(aabbId: getIndex());
    return [min[0], min[1]];
  }

  List<double> get max {
    final max = api.aabbMax(aabbId: getIndex());
    return [max[0], max[1]];
  }

  List<double> get center {
    final center = api.aabbCenter(aabbId: getIndex());
    return [center[0], center[1]];
  }

  double get width {
    return max[0] - min[0];
  }

  double get height {
    return max[1] - min[1];
  }

  List<double> get size {
    return [width, height];
  }

  // Methods
  bool drop() {
    return api.aabbDropBulk(aabbIds: [getIndex()])[0] == 1;
  }

  bool contains(double x, double y) {
    return api.aabbContainsPoint(
        aabbId: getIndex(), point: F64Array2(Float64List.fromList([x, y])));
  }

  bool containsAABB(AABB other) {
    return api.aabbContainsAabb(
        aabbLeftId: getIndex(), aabbRightId: other.getIndex());
  }

  bool intersectsAABB(AABB other) {
    return api.aabbIntersectsAabb(
        aabbLeftId: getIndex(), aabbRightId: other.getIndex());
  }

  AABB merge(AABB other) {
    final newId =
        api.aabbMerge(aabbLeftId: getIndex(), aabbRightId: other.getIndex());
    return AABB.fromIndex(newId);
  }

  void mergeWith(AABB other) {
    api.aabbMergeWith(aabb: getIndex(), other: other.getIndex());
  }
}

File: native/src/aabb.rs
use generational_arena::Index;

#[derive(Debug, Clone, Copy)]
pub struct AABB {
    pub id: Option<Index>,
    pub min_x: f64,
    pub min_y: f64,
    pub max_x: f64,
    pub max_y: f64,
}

impl AABB {
    pub fn empty() -> Self {
        Self {
            id: None,
            min_x: f64::INFINITY,
            min_y: f64::INFINITY,
            max_x: f64::NEG_INFINITY,
            max_y: f64::NEG_INFINITY,
        }
    }

    pub fn new(min_x: f64, min_y: f64, max_x: f64, max_y: f64) -> Self {
        Self {
            id: None,
            min_x,
            min_y,
            max_x,
            max_y,
        }
    }

    pub fn min(&self) -> [f64; 2] {
        [self.min_x, self.min_y]
    }

    pub fn max(&self) -> [f64; 2] {
        [self.max_x, self.max_y]
    }

    pub fn center(&self) -> [f64; 2] {
        let x = (self.max_x + self.min_x) / 2.0;
        let y = (self.max_y + self.min_y) / 2.0;
        [x, y]
    }

    pub fn size(&self) -> [f64; 2] {
        let width = self.max_x - self.min_x;
        let height = self.max_y - self.min_y;
        [width, height]
    }

    pub fn contains_point(&self, point_x: f64, point_y: f64) -> bool {
        point_x >= self.min_x
            && point_x <= self.max_x
            && point_y >= self.min_y
            && point_y <= self.max_y
    }

    pub fn intersects_aabb(&self, other: &Self) -> bool {
        self.max_x >= other.min_x
            && self.min_x <= other.max_x
            && self.max_y >= other.min_y
            && self.min_y <= other.max_y
    }

    pub fn merge(&self, other: &Self) -> Self {
        let min_x = self.min_x.min(other.min_x);
        let min_y = self.min_y.min(other.min_y);
        let max_x = self.max_x.max(other.max_x);
        let max_y = self.max_y.max(other.max_y);
        Self::new(min_x, min_y, max_x, max_y)
    }

    pub fn merge_with(&mut self, other: &Self) {
        self.min_x = self.min_x.min(other.min_x);
        self.min_y = self.min_y.min(other.min_y);
        self.max_x = self.max_x.max(other.max_x);
        self.max_y = self.max_y.max(other.max_y);
    }

    pub fn merge_point(&self, point_x: f64, point_y: f64) -> Self {
        let min_x = self.min_x.min(point_x);
        let min_y = self.min_y.min(point_y);
        let max_x = self.max_x.max(point_x);
        let max_y = self.max_y.max(point_y);
        Self::new(min_x, min_y, max_x, max_y)
    }

    pub fn merge_with_point(&mut self, point_x: f64, point_y: f64) {
        self.min_x = self.min_x.min(point_x);
        self.min_y = self.min_y.min(point_y);
        self.max_x = self.max_x.max(point_x);
        self.max_y = self.max_y.max(point_y);
    }

    pub fn contains_aabb(&self, other: &Self) -> bool {
        self.min_x <= other.min_x
            && self.min_y <= other.min_y
            && self.max_x >= other.max_x
            && self.max_y >= other.max_y
    }

    pub fn overlap_ratio(&self, other: &Self) -> f64 {
        let x_overlap = self.max_x.min(other.max_x) - self.min_x.max(other.min_x);
        let y_overlap = self.max_y.min(other.max_y) - self.min_y.max(other.min_y);

        if x_overlap > 0.0 && y_overlap > 0.0 {
            let overlap_area = x_overlap * y_overlap;
            let self_area = (self.max_x - self.min_x) * (self.max_y - self.min_y);
            let other_area = (other.max_x - other.min_x) * (other.max_y - other.min_y);
            overlap_area / (self_area + other_area - overlap_area)
        } else {
            0.0
        }
    }

    pub fn area(&self) -> f64 {
        let size = self.size();
        size[0] * size[1]
    }

    pub fn longest_axis(&self) -> usize {
        let size = self.size();
        if size[0] > size[1] {
            0
        } else {
            1
        }
    }
}
